gluon-l3roamd: site: make node_client_prefix6 optional+deprecated

T-X opened the pull request:
The site.node_client_prefix6() is only used internally by the l3roamd
protocol. Therefore it is unnecessary to expose it to an administrator.
    
Instead, if node_client_prefix6 is unspecified in the site, generate an
IPv6 Unique Local Address prefix from the site domain_seed.
    
This updates the site documentation as well and marks this setting as
both optional and deprecated.
    
Note: If you had the node_client_prefix6 specified before and want to
use the new autogeneration from the domain_seed instead then this will
break compatibility and will need a gluon-scheduled-domain switch.
    
Signed-off-by: Linus Lüssing \<linus.luessing@c0d3.blue\>

T-X commented:
Commit is rebased on top of #2553.

Tested that the node_client_prefix6() is both correct in the l3roamd command line and in the babel config file. Still ToDo: Test roaming of a client device with the autogenerated prefix.

@mkg20001 @christf

mkg20001 commented on package/gluon-mesh-babel/check_site.lua:
```diff
@@ -1,4 +1,5 @@
-need_string_match(in_domain({'node_client_prefix6'}), '^[%x:]+/64$')
+need_string_match(in_domain({'prefix6'}), '^[%x:]+/64$')
+need_string_match(in_domain({'node_prefix6'}), '^[%x:]+/64$')
```
this is already required by layer3-common

T-X replied:
I think I'd add that anyways, even if a dependency pulls in the same check, too. Tracking if a check was added by another dependency seems tedious. Also if someone were to remove the usage of the site setting for gluon-layer3-common later, there is a high chance that the check would be removed in gluon-layer3-common, too and that the needed check in gluon-mesh-babel would then be missed.

Therefore it seems the most simple and less error prone approach to me, to stick to this simple rule: If this package requires a site/domain setting, then add it to this package's check_site.lua. Even if another dependcy package has it in its check_site.lua as well.

mkg20001 commented on package/gluon-mesh-babel/check_site.lua:
```diff
@@ -1,4 +1,5 @@
-need_string_match(in_domain({'node_client_prefix6'}), '^[%x:]+/64$')
+need_string_match(in_domain({'prefix6'}), '^[%x:]+/64$')
```
this is already required by gluon-core

T-X replied:
same reasoning as above

mkg20001 reviewed (commented):
not sure if we really need to have a unique l3roamd prefix. it might make sense to have it domain-specific aswell though then, which maybe would make handling of accidental cross-domain meshing better.

mkg20001 commented on package/gluon-mesh-babel/luasrc/lib/gluon/upgrade/300-gluon-mesh-babel-mkconfig:
```diff
@@ -14,7 +15,7 @@ file:write("import-table 254\n")
 file:write("redistribute ip " .. site.next_node.ip6() .. "/128 deny\n")
 file:write("redistribute ip " .. site.prefix6() .. " eq 128  allow\n")
-file:write("redistribute ip " .. site.node_client_prefix6() .. " eq 128  allow\n")
+file:write("redistribute ip " .. l3.node_client_prefix6() .. " eq 128  allow\n")
```
I don't know if this is correct at all, since the prefix per doc doesn't seem to be used for routing packets? Or is it being used for that in some kindof crazy way?

T-X replied:
As far as I understand l3roamd uses this prefix to announce the roaming of a client to another node. The receiving node listens on an address in this range with IP_FREEBIND. To be able to receive such an announcement, nodes between the old and the new node for the client that roamed need to forward these l3roamd packets with this unicast node-client-IP destination address.

christf replied:
this is exactly how it works.

mkg20001 commented on package/gluon-layer3-common/luasrc/usr/lib/lua/gluon/l3.lua:
```diff
@@ -0,0 +1,21 @@
+-- returns a prefix generated from the domain-seed
+-- for l3roamd -P <node-client-prefix>
+function M.node_client_prefix6()
+	local key = "gluon-l3roamd.node_client_prefix6"
```
This key could include the current domain, for the client address to be unique per domain. Would that be useful?

T-X replied:
I think that's what should already be happening a few lines later, with the call to `util.domain_seed_bytes(key, 7)`. The key + the site's/domain's domain_seed will be the seeds to generate a prefix unique for this domain.

Check the other usages of domain_seed_bytes() in Gluon. We are using it like `...domain_seed_bytes("gluon-mesh-vxlan", 3)' to generate a VXLAN ID which is unique per domain, as well.

mkg20001 replied:
Ah, perfect

mkg20001 commented:
@T-X the firewall pr is now merged, could you rebase this then? Other than that LGTM

T-X commented:
Changelog v2:

* rebased to current master (renamed gluon-layer3-common to gluon-mesh-layer3-common)
* added node_client_prefix6 to gluon-mesh-layer3-common's check_site.lua, with required=false
* removed additions of prefix6 and node_prefix6 in check_site.lua again

mkg20001 commented:
I think this is good for merge.

mkg20001 reviewed (approved):
LGTM

blocktrron commented on package/gluon-mesh-layer3-common/luasrc/usr/lib/lua/gluon/l3.lua:
```diff
@@ -0,0 +1,21 @@
+	local key = "gluon-l3roamd.node_client_prefix6"
+	local prefix = site.node_client_prefix6()
+
+	if not prefix
+	then
```
we usually put this in a single line.

AiyionPrime replied:
resolved in [b3a9221](https://github.com/freifunk-gluon/gluon/commit/b3a9221b07f083d95aff1b3057ccf5dafa7efd42)

blocktrron reviewed (approved):
LGTM apart from the minor style improvement.

AiyionPrime reviewed (approved):
Will merge this when the actions ran through.
